### PR TITLE
Initial set of Azure secrets for #539

### PR DESCRIPTION
**Pull request by dvasdekis**
### Description:
Please find in this PR an almost-complete set of Azure secrets as [sourced from here](https://www.powershellgallery.com/packages/AzSK.AzureDevOps/0.9.8/Content/Framework%5CConfigurations%5CSVT%5CAzureDevOps%5CCredentialPatterns.xml), plus some new ones from my own experience.

### Checklist:

* [x] Does your PR pass tests?
* [x] Have you written new tests for your changes?
* [x] Have you lint your code locally prior to submission?

### CODE REVIEW

**Comment by dvasdekis** [comment]
I've commented out the secret types I couldn't get to work (mostly the XML ones) but the rest pass `go run main.go` now. Hopefully you can approve and pull in 👍

File: .gitignore [hunk 1]
```diff
@@ -18,3 +18,12 @@ cmd/generate/config/gitleaks.toml
...
 dist/
+
+# Hugo build files (created automatically by VSCode Go extension)
+.hugo_build.lock
+
+# Hugo Sitemap.xml files (created automatically by VSCode Go extension)
+public/index.xml
+public/sitemap.xml
+public/categories/index.xml
+public/tags/index.xml
```
Comment (jessehouwing) [inline .gitignore:29]:
```suggestion
```

These should probably not be part of the PR.

Comment (dvasdekis) [reply .gitignore:29]: These files are created by `go run main.go`. 

Should these `.xml` files be included in the repository? If they should, then I'll commit them. If not, they should be in the .gitignore. Given that they're automatically built, `.gitignore` seems more appropriate. Yes, it isn't an atomic PR, but they should be there.

Comment (nikolamalesevic) [reply .gitignore:29]: @dvasdekis Maybe leave this out and get the PR approved? I'm really looking forward for your changes in the next release. :)

**Comment by dvasdekis** [comment]
Sure, I've accepted the changes

**Comment by dvasdekis** [comment]
Hi, can I get these changed merged please?

**Comment by JoostVoskuil** [comment]
Hi @zricethezav , can this be merged? This is a big addition for the Microsoft techstack. Or is there an additional step required?

File: cmd/generate/config/rules/azure.go [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+	return validate(r, tps, nil)
+}
+
+func AzureStorageCredentialXStore() *config.Rule {
+	// define rule
+	r := config.Rule{
+		Description: "CSCAN0100 - Found Azure storage credential in source code file.",
+		RuleID:      "azure-storage-credential-xstore",
+		SecretGroup: 1,
+		Regex: generateUniqueTokenRegex(`[<XstoreAccountInfo].*accountSharedKey\s*=\s*['"].*['"]`),
```
Comment (rgmz) [inline cmd/generate/config/rules/azure.go:132]: Just curious, what's the reason for changing some of the patterns?

```diff
-<XstoreAccountInfo[ -~"\s\S\n\r\t]+accountSharedKey\s*=\s*"[^"]{30}[ -~"\s\S\n\r\t]+/>
+[<XstoreAccountInfo].*accountSharedKey\s*=\s*['"].*['"]
```

In particular, this won't work as expected because `[<XstoreAccountInfo]` will only match a single character from inside the square brackets, not `<XstoreAccountInfo`.

**Comment by tgrnwd** [comment]
Definitely interested in seeing these rules included. Anything I can do to help @dvasdekis ?

**Comment by dvasdekis** [comment]
These rules are complete. Not sure why they're not merged

**Comment by JoostVoskuil** [comment]
@zricethezav how can we complete this?

**Comment by zricethezav** [comment]
Actually, jk, there are some issues here. Most of these rules do not have `keywords`, which are necessary if gitleaks is to maintain reasonable performance. Opening up a follow up commit

**Comment by dvasdekis** [comment]
Okay, I can amend the original code with some keywords, when they're available from Azure. But for those in the spec that don't have keywords, are you planning to never merge then?

**Comment by nikolamalesevic** [comment]
@zricethezav This is not a way to handle this. Many of us have been waiting for this PR for a long time, with inexplicable months of silence from your side, and now jokes and fun and let's throw everything in the fire.

There must be a better way to communicate a way forward.

**Comment by JoostVoskuil** [comment]
@nikolamalesevic be kind please

**Comment by zricethezav** [comment]
@nikolamalesevic lol I'm really really sorry for not providing good enough free labor. get real dude

Anyways, this PR adds rules to the _default configuration_. If you want to use these rules for your own scans then _you can_, you just need to specify it using [`--config`](https://twitter.com/dog_rates/status/1669469025762684928) or one of the other methods. 

I admit, I was trigger happy and merged this PR too quickly, something that should be added is a check that makes sure all new rules have at least one keyword and that the delta in scan time does not exceed some %.

**Comment by zricethezav** [comment]
@dvasdekis 
> Okay, I can amend the original code with some keywords, when they're available from Azure. But for those in the spec that don't have keywords, are you planning to never merge then?

I'd be happy to merge if the rules can be updated with keywords. The reason why keywords are so important is because gitleaks does a string compare on a chunk of text before attempting to match a regex. So if no keyword is specified, then that rule would be running regex on _every_ chunk of text, which slows things down dramatically. You can see how this is implemented here: https://github.com/gitleaks/gitleaks/blob/7804d652c0460f0b61ea9bb2c1f202bfcdbbf144/detect/detect.go#L578-L596

Feel free to reopen a PR when keywords have been added. Again, sorry for the merging then reverting.

File: cmd/generate/config/rules/azure.go (hunk 2 above)
Comment (dvasdekis) [reply cmd/generate/config/rules/azure.go:132]: Probably because there were like 100 rules and I blew through them with all the enthusiasm that I had, now exhausted. When I get some free time I'll try it again, with the above change

**Comment by zricethezav** [comment]
> The comment was meant to be an honest feedback specific to handling of this PR (however negative), and never did I say that I do not appreciate a hard work put into this. I do apologize if I came out as being rude.

It's all good, I need a performance check to make sure new rules don't increase the scan time by some %. Again, I'm happy to review new rules if someone is up to the task.
